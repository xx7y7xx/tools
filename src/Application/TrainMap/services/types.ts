export interface TrainPosition {
  id: string;
  trainNumber: string;
  pocsag1234002Data: {
    pocsagMsgTimestamp: string;
    wgs84_latitude?: number;
    wgs84_longitude?: number;
    gcj02_latitude?: number;
    gcj02_longitude?: number;
    rawMessage?: string;
  };
}

export interface WebSocketMessage {
  type: 'train_position' | 'train_positions' | 'error' | 'ping' | 'pong';
  data?: any;
  timestamp: string;
}

export interface TrainPositionMessage extends WebSocketMessage {
  type: 'train_position';
  data: TrainPosition;
}

export interface TrainPositionsMessage extends WebSocketMessage {
  type: 'train_positions';
  data: TrainPosition[];
}
