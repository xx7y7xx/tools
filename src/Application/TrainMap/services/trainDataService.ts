import { TrainPosition, WebSocketMessage } from './types';

export class TrainDataService {
  private ws: WebSocket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private updateCallbacks: ((train: TrainPosition) => void)[] = [];
  private allTrainsCallbacks: ((trains: TrainPosition[]) => void)[] = [];
  private connectionCallbacks: ((connected: boolean) => void)[] = [];
  private isConnected = false;

  constructor(private serverUrl: string = 'ws://localhost:8080/trains') {}

  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.serverUrl);

        this.ws.onopen = () => {
          console.log('🔌 Connected to Train WebSocket Server');
          this.isConnected = true;
          this.reconnectAttempts = 0;
          this.notifyConnectionStatus(true);
          resolve();
        };

        this.ws.onmessage = (event) => {
          try {
            const message: WebSocketMessage = JSON.parse(event.data);
            this.handleMessage(message);
          } catch (error) {
            console.error('❌ Error parsing WebSocket message:', error);
          }
        };

        this.ws.onclose = (event) => {
          console.log(
            '🔌 WebSocket connection closed:',
            event.code,
            event.reason
          );
          this.isConnected = false;
          this.notifyConnectionStatus(false);

          if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.scheduleReconnect();
          }
        };

        this.ws.onerror = (error) => {
          console.error('❌ WebSocket error:', error);
          reject(error);
        };
      } catch (error) {
        console.error('❌ Error creating WebSocket connection:', error);
        reject(error);
      }
    });
  }

  private handleMessage(message: WebSocketMessage) {
    switch (message.type) {
      case 'train_position':
        if (message.data) {
          this.notifyTrainUpdate(message.data);
        }
        break;

      case 'train_positions':
        if (message.data) {
          this.notifyAllTrainsUpdate(message.data);
        }
        break;

      case 'pong':
        console.log('📡 Server ping response received');
        break;

      case 'error':
        console.error('❌ Server error:', message.data);
        break;

      default:
        console.log('❓ Unknown message type:', message.type);
    }
  }

  private scheduleReconnect() {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

    console.log(
      `🔄 Attempting to reconnect in ${delay}ms (attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts})`
    );

    setTimeout(() => {
      if (!this.isConnected) {
        this.connect().catch((error) => {
          console.error('❌ Reconnection failed:', error);
        });
      }
    }, delay);
  }

  disconnect() {
    if (this.ws) {
      this.ws.close(1000, 'Client disconnect');
      this.ws = null;
    }
    this.isConnected = false;
    this.notifyConnectionStatus(false);
  }

  // Subscribe to specific train updates
  subscribeToTrain(trainId: string) {
    if (this.ws && this.isConnected) {
      this.ws.send(
        JSON.stringify({
          type: 'subscribe_train',
          trainId,
        })
      );
    }
  }

  // Unsubscribe from specific train updates
  unsubscribeFromTrain(trainId: string) {
    if (this.ws && this.isConnected) {
      this.ws.send(
        JSON.stringify({
          type: 'unsubscribe_train',
          trainId,
        })
      );
    }
  }

  // Request all train positions
  requestAllTrains() {
    if (this.ws && this.isConnected) {
      this.ws.send(
        JSON.stringify({
          type: 'get_all_trains',
        })
      );
    }
  }

  // Send ping to keep connection alive
  ping() {
    if (this.ws && this.isConnected) {
      this.ws.send(
        JSON.stringify({
          type: 'ping',
          timestamp: Date.now(),
        })
      );
    }
  }

  // Event listeners
  onTrainUpdate(callback: (train: TrainPosition) => void): () => void {
    this.updateCallbacks.push(callback);
    return () => {
      const index = this.updateCallbacks.indexOf(callback);
      if (index > -1) {
        this.updateCallbacks.splice(index, 1);
      }
    };
  }

  onAllTrainsUpdate(callback: (trains: TrainPosition[]) => void): () => void {
    this.allTrainsCallbacks.push(callback);
    return () => {
      const index = this.allTrainsCallbacks.indexOf(callback);
      if (index > -1) {
        this.allTrainsCallbacks.splice(index, 1);
      }
    };
  }

  onConnectionChange(callback: (connected: boolean) => void): () => void {
    this.connectionCallbacks.push(callback);
    return () => {
      const index = this.connectionCallbacks.indexOf(callback);
      if (index > -1) {
        this.connectionCallbacks.splice(index, 1);
      }
    };
  }

  private notifyTrainUpdate(train: TrainPosition) {
    this.updateCallbacks.forEach((callback) => callback(train));
  }

  private notifyAllTrainsUpdate(trains: TrainPosition[]) {
    this.allTrainsCallbacks.forEach((callback) => callback(trains));
  }

  private notifyConnectionStatus(connected: boolean) {
    this.connectionCallbacks.forEach((callback) => callback(connected));
  }

  // Get connection status
  getConnectionStatus(): boolean {
    return this.isConnected;
  }

  // Start automatic ping
  startAutoPing(interval: number = 30000) {
    setInterval(() => {
      this.ping();
    }, interval);
  }
}

// Export singleton instance
export const trainDataService = new TrainDataService();
