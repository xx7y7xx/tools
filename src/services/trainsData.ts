import axios, { AxiosResponse } from 'axios';

// Configuration for the trains-data API
const TRAINS_DATA_BASE_URL = 'http://localhost:3000/trains-data';

// Interface for train data from checiMap.json
export interface TrainInfo {
  from_station: string;
  station_train_code: string;
  to_station: string;
  total_num: string;
  train_no: string;
}

// Interface for the complete trains data response
export interface TrainsDataResponse {
  [trainCode: string]: TrainInfo;
}

/**
 * Fetches trains data from checiMap.json for a specific date
 * @param dateString - Date in YYYYMMDD format (e.g., '20250609'). Required since there's no root checiMap.json
 * @returns Promise containing trains data
 */
export const fetchTrainsData = async (
  dateString: string
): Promise<TrainsDataResponse> => {
  try {
    const response: AxiosResponse = await axios.get(
      `${TRAINS_DATA_BASE_URL}/${dateString}/checiMap.json`,
      {
        headers: {
          Accept: 'application/json',
        },
      }
    );

    return response.data;
  } catch (error) {
    console.error('Error fetching trains data:', error);
    throw new Error(
      `Failed to fetch trains data for ${dateString}: ${
        error instanceof Error ? error.message : 'Unknown error'
      }`
    );
  }
};

/**
 * Fetches data for a specific train across multiple dates
 * @param trainCode - Train code (e.g., '1461')
 * @param dates - Array of dates in YYYYMMDD format
 * @returns Promise containing train data trends
 */
export const fetchTrainTrends = async (
  trainCode: string,
  dates: string[]
): Promise<{ date: string; total_num: number; trainInfo?: TrainInfo }[]> => {
  const results = await Promise.allSettled(
    dates.map(async (date) => {
      try {
        const data = await fetchTrainsData(date);
        const trainInfo = data[trainCode];
        return {
          date,
          total_num: trainInfo ? parseInt(trainInfo.total_num) : 0,
          trainInfo: trainInfo || undefined,
        };
      } catch (error) {
        return {
          date,
          total_num: 0,
          trainInfo: undefined,
        };
      }
    })
  );

  return results
    .filter(
      (result): result is PromiseFulfilledResult<any> =>
        result.status === 'fulfilled'
    )
    .map((result) => result.value)
    .sort((a, b) => a.date.localeCompare(b.date));
};
