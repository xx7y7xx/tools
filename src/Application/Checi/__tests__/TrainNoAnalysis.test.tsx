// All tests for TrainNoAnalysis have been temporarily removed for step-by-step error fixing.

import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import TrainNoAnalysis from '../TrainNoAnalysis';
import { fetchAllHistoricalData } from '../../../services/trainsData';

// Mock Ant Design's responsive observer to prevent addListener error
jest.mock('antd/lib/_util/responsiveObserver', () => ({
  __esModule: true,
  default: () => ({
    subscribe: jest.fn(),
    unsubscribe: jest.fn(),
  }),
}));

jest.mock('../../../services/trainsData', () => ({
  fetchAllHistoricalData: jest.fn(),
}));

describe('TrainNoAnalysis (happy path)', () => {
  it('renders statistics when data loads', async () => {
    // Mock data for two dates and two trains
    (fetchAllHistoricalData as jest.Mock).mockResolvedValue({
      '20241120': {
        G1: {
          train_no: 'T100',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
        G2: {
          train_no: 'T200',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
      },
      '20241121': {
        G1: {
          train_no: 'T100',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
        G2: {
          train_no: 'T200',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
      },
    });
    const onError = jest.fn();
    const onLoadingChange = jest.fn();
    render(
      <TrainNoAnalysis onError={onError} onLoadingChange={onLoadingChange} />
    );
    // Check for statistic titles
    expect(await screen.findByText('总车次数')).toBeInTheDocument();
    expect(await screen.findByText('有变化车次')).toBeInTheDocument();
    expect(await screen.findByText('无变化车次')).toBeInTheDocument();
    expect(await screen.findByText('变化率')).toBeInTheDocument();
    // Check for the numbers (should be at least two '2's and two '0's)
    expect((await screen.findAllByText('2')).length).toBeGreaterThanOrEqual(2);
    expect((await screen.findAllByText('0')).length).toBeGreaterThanOrEqual(2);
  });
});

describe('TrainNoAnalysis (snapshot)', () => {
  it('matches snapshot', async () => {
    // Mock data for two dates and two trains
    (fetchAllHistoricalData as jest.Mock).mockResolvedValue({
      '20241120': {
        G1: {
          train_no: 'T100',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
        G2: {
          train_no: 'T200',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
      },
      '20241121': {
        G1: {
          train_no: 'T100',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
        G2: {
          train_no: 'T200',
          from_station: '',
          to_station: '',
          station_train_code: '',
          total_num: '1',
        },
      },
    });
    const onError = jest.fn();
    const onLoadingChange = jest.fn();
    const { container } = render(
      <TrainNoAnalysis onError={onError} onLoadingChange={onLoadingChange} />
    );

    // Wait for the component to finish loading
    await screen.findByText('总车次数');

    // Create snapshot
    expect(container).toMatchSnapshot();
  });
});
