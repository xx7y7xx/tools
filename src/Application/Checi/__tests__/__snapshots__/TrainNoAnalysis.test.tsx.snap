// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrainNoAnalysis (snapshot) matches snapshot 1`] = `
<div>
  <div
    style="padding: 16px;"
  >
    <h2>
      train_no变化分析
    </h2>
    <p
      style="color: rgb(102, 102, 102); margin-bottom: 24px;"
    >
      分析所有车次在不同日期间的train_no字段是否发生变化
    </p>
    <div
      class="ant-row css-dev-only-do-not-override-1uq9j6g"
      style="margin-left: -8px; margin-right: -8px; margin-bottom: 24px;"
    >
      <div
        class="ant-col ant-col-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                总车次数
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(24, 144, 255);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    2
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                有变化车次
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(245, 34, 45);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    0
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                无变化车次
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(82, 196, 26);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    2
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                变化率
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(114, 46, 209);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    0
                  </span>
                  <span
                    class="ant-statistic-content-value-decimal"
                  >
                    .0
                  </span>
                </span>
                <span
                  class="ant-statistic-content-suffix"
                >
                  %
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-row css-dev-only-do-not-override-1uq9j6g"
      style="margin-left: -8px; margin-right: -8px; margin-bottom: 24px;"
    >
      <div
        class="ant-col ant-col-12 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                最大变化数
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(250, 140, 22);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    1
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-12 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-1uq9j6g"
            >
              <div
                class="ant-statistic-title"
              >
                平均变化数
              </div>
              <div
                class="ant-statistic-content"
                style="color: rgb(19, 194, 194);"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    1
                  </span>
                  <span
                    class="ant-statistic-content-value-decimal"
                  >
                    .00
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
      style="margin-bottom: 24px;"
    >
      <div
        class="ant-card-head"
      >
        <div
          class="ant-card-head-wrapper"
        >
          <div
            class="ant-card-head-title"
          >
            变化分布图
          </div>
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <canvas
          height="150"
          role="img"
          width="300"
        />
      </div>
    </div>
    <div
      class="ant-card ant-card-bordered css-dev-only-do-not-override-1uq9j6g"
    >
      <div
        class="ant-card-head"
      >
        <div
          class="ant-card-head-wrapper"
        >
          <div
            class="ant-card-head-title"
          >
            详细分析结果
          </div>
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <div
          class="ant-table-wrapper css-dev-only-do-not-override-1uq9j6g"
        >
          <div
            class="ant-spin-nested-loading css-dev-only-do-not-override-1uq9j6g"
          >
            <div
              class="ant-spin-container"
            >
              <div
                class="ant-table css-dev-only-do-not-override-1uq9j6g ant-table-scroll-horizontal"
              >
                <div
                  class="ant-table-container"
                >
                  <div
                    class="ant-table-content"
                    style="overflow-x: auto; overflow-y: hidden;"
                  >
                    <table
                      style="width: 800px; min-width: 100%; table-layout: auto;"
                    >
                      <colgroup>
                        <col
                          style="width: 120px;"
                        />
                        <col
                          style="width: 120px;"
                        />
                        <col
                          style="width: 100px;"
                        />
                        <col />
                        <col
                          style="width: 100px;"
                        />
                      </colgroup>
                      <thead
                        class="ant-table-thead"
                      >
                        <tr>
                          <th
                            aria-label="车次代码"
                            class="ant-table-cell ant-table-column-has-sorters"
                            scope="col"
                            tabindex="0"
                          >
                            <div
                              class="ant-table-column-sorters"
                            >
                              <span
                                class="ant-table-column-title"
                              >
                                车次代码
                              </span>
                              <span
                                class="ant-table-column-sorter ant-table-column-sorter-full"
                              >
                                <span
                                  aria-hidden="true"
                                  class="ant-table-column-sorter-inner"
                                >
                                  <span
                                    aria-label="caret-up"
                                    class="anticon anticon-caret-up ant-table-column-sorter-up"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-up"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                  <span
                                    aria-label="caret-down"
                                    class="anticon anticon-caret-down ant-table-column-sorter-down"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            <div
                              class="ant-table-filter-column"
                            >
                              <span
                                class="ant-table-column-title"
                              >
                                是否有变化
                              </span>
                              <span
                                class="ant-dropdown-trigger ant-table-filter-trigger"
                                role="button"
                                tabindex="-1"
                              >
                                <span
                                  aria-label="filter"
                                  class="anticon anticon-filter"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="filter"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="64 64 896 896"
                                    width="1em"
                                  >
                                    <path
                                      d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </div>
                          </th>
                          <th
                            aria-label="变化数量"
                            class="ant-table-cell ant-table-column-has-sorters"
                            scope="col"
                            tabindex="0"
                          >
                            <div
                              class="ant-table-column-sorters"
                            >
                              <span
                                class="ant-table-column-title"
                              >
                                变化数量
                              </span>
                              <span
                                class="ant-table-column-sorter ant-table-column-sorter-full"
                              >
                                <span
                                  aria-hidden="true"
                                  class="ant-table-column-sorter-inner"
                                >
                                  <span
                                    aria-label="caret-up"
                                    class="anticon anticon-caret-up ant-table-column-sorter-up"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-up"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                  <span
                                    aria-label="caret-down"
                                    class="anticon anticon-caret-down ant-table-column-sorter-down"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            所有train_no值
                          </th>
                          <th
                            aria-label="出现日期数"
                            class="ant-table-cell ant-table-column-has-sorters"
                            scope="col"
                            tabindex="0"
                          >
                            <div
                              class="ant-table-column-sorters"
                            >
                              <span
                                class="ant-table-column-title"
                              >
                                出现日期数
                              </span>
                              <span
                                class="ant-table-column-sorter ant-table-column-sorter-full"
                              >
                                <span
                                  aria-hidden="true"
                                  class="ant-table-column-sorter-inner"
                                >
                                  <span
                                    aria-label="caret-up"
                                    class="anticon anticon-caret-up ant-table-column-sorter-up"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-up"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                  <span
                                    aria-label="caret-down"
                                    class="anticon anticon-caret-down ant-table-column-sorter-down"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="ant-table-tbody"
                      >
                        <tr
                          aria-hidden="true"
                          class="ant-table-measure-row"
                          style="height: 0px; font-size: 0px;"
                        >
                          <td
                            style="padding: 0px; border: 0px; height: 0px;"
                          >
                            <div
                              style="height: 0px; overflow: hidden;"
                            >
                               
                            </div>
                          </td>
                          <td
                            style="padding: 0px; border: 0px; height: 0px;"
                          >
                            <div
                              style="height: 0px; overflow: hidden;"
                            >
                               
                            </div>
                          </td>
                          <td
                            style="padding: 0px; border: 0px; height: 0px;"
                          >
                            <div
                              style="height: 0px; overflow: hidden;"
                            >
                               
                            </div>
                          </td>
                          <td
                            style="padding: 0px; border: 0px; height: 0px;"
                          >
                            <div
                              style="height: 0px; overflow: hidden;"
                            >
                               
                            </div>
                          </td>
                          <td
                            style="padding: 0px; border: 0px; height: 0px;"
                          >
                            <div
                              style="height: 0px; overflow: hidden;"
                            >
                               
                            </div>
                          </td>
                        </tr>
                        <tr
                          class="ant-table-row ant-table-row-level-0"
                          data-row-key="G1"
                        >
                          <td
                            class="ant-table-cell"
                          >
                            G1
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            <span
                              class="ant-tag ant-tag-green css-dev-only-do-not-override-1uq9j6g"
                            >
                              无变化
                            </span>
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            1
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            <div>
                              <span
                                class="ant-tag ant-tag-blue css-dev-only-do-not-override-1uq9j6g"
                                style="margin-bottom: 4px;"
                              >
                                T100
                              </span>
                            </div>
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            2
                          </td>
                        </tr>
                        <tr
                          class="ant-table-row ant-table-row-level-0"
                          data-row-key="G2"
                        >
                          <td
                            class="ant-table-cell"
                          >
                            G2
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            <span
                              class="ant-tag ant-tag-green css-dev-only-do-not-override-1uq9j6g"
                            >
                              无变化
                            </span>
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            1
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            <div>
                              <span
                                class="ant-tag ant-tag-blue css-dev-only-do-not-override-1uq9j6g"
                                style="margin-bottom: 4px;"
                              >
                                T200
                              </span>
                            </div>
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            2
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
              <ul
                class="ant-pagination ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-1uq9j6g"
              >
                <li
                  class="ant-pagination-total-text"
                >
                  第 1-2 条，共 2 条记录
                </li>
                <li
                  aria-disabled="true"
                  class="ant-pagination-prev ant-pagination-disabled"
                  title="Previous Page"
                >
                  <button
                    class="ant-pagination-item-link"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      aria-label="left"
                      class="anticon anticon-left"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="left"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                        />
                      </svg>
                    </span>
                  </button>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                  tabindex="0"
                  title="1"
                >
                  <a
                    rel="nofollow"
                  >
                    1
                  </a>
                </li>
                <li
                  aria-disabled="true"
                  class="ant-pagination-next ant-pagination-disabled"
                  title="Next Page"
                >
                  <button
                    class="ant-pagination-item-link"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      aria-label="right"
                      class="anticon anticon-right"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="right"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                        />
                      </svg>
                    </span>
                  </button>
                </li>
                <li
                  class="ant-pagination-options"
                >
                  <div
                    aria-label="Page Size"
                    class="ant-select ant-select-outlined ant-pagination-options-size-changer css-dev-only-do-not-override-1uq9j6g ant-select-single ant-select-show-arrow ant-select-show-search"
                  >
                    <div
                      class="ant-select-selector"
                    >
                      <span
                        class="ant-select-selection-search"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-controls="undefined_list"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          aria-label="Page Size"
                          aria-owns="undefined_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          role="combobox"
                          type="search"
                          value=""
                        />
                      </span>
                      <span
                        class="ant-select-selection-item"
                        title="20 / page"
                      >
                        20 / page
                      </span>
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-arrow"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-select-suffix"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
